File: src/lib/kobo.ts
export interface KoboAttachment {
  uid: string;
  mimetype: string;
  filename: string;
  media_file_basename: string;
  question_xpath: string;
  is_deleted: boolean;
  download_url: string;
  download_large_url: string;
  download_medium_url: string;
  download_small_url: string;
}

export interface KoboSubmission {
  _id: number;
  _uuid: string;
  _xform_id_string: string;
  _submission_time: string;
  _submitted_by: string | null;
  _attachments: KoboAttachment[];
  [key: string]: unknown;
}

/**
 * Returns true if the given URL's hostname matches the hostname of baseUrl.
 * Used to prevent SSRF on the media proxy.
 */
export function isAllowedMediaHost(url: string, baseUrl: string): boolean {
  try {
    const target = new URL(url);
    const base = new URL(baseUrl);
    return target.hostname === base.hostname;
  } catch {
    return false;
  }
}

/**
 * Returns only non-deleted image attachments from a submission.
 */
export function imageAttachments(submission: KoboSubmission): KoboAttachment[] {
  return (submission._attachments ?? []).filter(
    (a) => !a.is_deleted && a.mimetype.startsWith("image/")
  );
}

/**
 * Returns only non-deleted non-image attachments (files, audio, video, etc.)
 * that are referenced (i.e. their basename appears as a string value) in the
 * given reference object (defaults to the full submission if omitted).
 */
export function nonImageAttachments(
  submission: KoboSubmission,
  reference?: Record<string, unknown>
): KoboAttachment[] {
  const filenames = submissionImageFilenames(
    reference !== undefined ? { ...submission, ...reference, _attachments: submission._attachments } : submission
  );
  return (submission._attachments ?? []).filter(
    (a) => !a.is_deleted && !a.mimetype.startsWith("image/") && filenames.has(a.media_file_basename)
  );
}

/**
 * Returns only non-deleted non-image attachments (files, audio, video, etc.)
 */
export function fileAttachments(submission: KoboSubmission): KoboAttachment[] {
  return (submission._attachments ?? []).filter(
    (a) => !a.is_deleted && !a.mimetype.startsWith("image/")
  );
}

/**
 * Recursively collects all string values from a value tree (objects, arrays, primitives).
 * Skips the _attachments key to avoid scanning attachment metadata as filenames.
 */
function collectStringValues(value: unknown, names: Set<string>): void {
  if (typeof value === "string") {
    if (value.length > 0) names.add(value);
  } else if (Array.isArray(value)) {
    for (const item of value) collectStringValues(item, names);
  } else if (value !== null && typeof value === "object") {
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      if (k === "_attachments") continue;
      collectStringValues(v, names);
    }
  }
}

/**
 * Collects all string values from the submission JSON (excluding _attachments),
 * including values nested inside repeat group arrays.
 * Used to identify which attachments are actually referenced in the submission data —
 * the REST service may filter out some questions, so only referenced images are forwarded.
 */
export function submissionImageFilenames(submission: KoboSubmission): Set<string> {
  const names = new Set<string>();
  for (const [key, value] of Object.entries(submission)) {
    if (key === "_attachments") continue;
    collectStringValues(value, names);
  }
  return names;
}

/**
 * Returns non-deleted attachments (all mimetypes) whose media_file_basename
 * appears as a string value in the reference payload.
 * When a fields subset is active, pass the filtered jsonPayload as reference
 * so only attachments relevant to those fields are fetched and forwarded.
 * Handles repeat group arrays by scanning recursively.
 */
export function attachmentsToForward(
  submission: KoboSubmission,
  reference?: Record<string, unknown>
): KoboAttachment[] {
  const scanTarget = reference ?? submission;
  const filenames = new Set<string>();
  for (const [key, value] of Object.entries(scanTarget)) {
    if (key === "_attachments") continue;
    collectStringValues(value, filenames);
  }
  return (submission._attachments ?? []).filter(
    (a) => !a.is_deleted && filenames.has(a.media_file_basename)
  );
}

/** @deprecated Use attachmentsToForward instead */
export function imageAttachmentsToForward(submission: KoboSubmission): KoboAttachment[] {
  return attachmentsToForward(submission);
}

/**
 * Format a Kobo submission time string for display.
 */
export function formatSubmissionTime(iso: string): string {
  try {
    return new Date(iso).toLocaleString("en-GB", {
      day: "2-digit",
      month: "short",
      year: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  } catch {
    return iso;
  }
}

File: app/lib/kobo.py
"""
KoboToolbox attachment helpers and SSRF guard.
Port of src/lib/kobo.ts.
"""

from urllib.parse import urlparse


def is_allowed_media_host(url: str, base_url: str) -> bool:
    """Returns True if the URL's hostname matches the base_url hostname (SSRF guard)."""
    try:
        return urlparse(url).hostname == urlparse(base_url).hostname
    except Exception:
        return False


def image_attachments(submission: dict) -> list[dict]:
    return [
        a for a in (submission.get('_attachments') or [])
        if not a.get('is_deleted') and (a.get('mimetype') or '').startswith('image/')
    ]


def file_attachments(submission: dict) -> list[dict]:
    return [
        a for a in (submission.get('_attachments') or [])
        if not a.get('is_deleted') and not (a.get('mimetype') or '').startswith('image/')
    ]


def _submission_string_values(submission: dict) -> set[str]:
    names: set[str] = set()
    for key, value in submission.items():
        if key == '_attachments':
            continue
        if isinstance(value, str) and value:
            names.add(value)
    return names


def attachments_to_forward(submission: dict, reference: dict | None = None) -> list[dict]:
    """
    Returns non-deleted attachments whose media_file_basename appears as a
    string value in the reference payload (or full submission if reference is None).
    """
    scan_target = reference if reference is not None else submission
    filenames: set[str] = set()
    for key, value in scan_target.items():
        if key == '_attachments':
            continue
        if isinstance(value, str) and value:
            filenames.add(value)
    return [
        a for a in (submission.get('_attachments') or [])
        if not a.get('is_deleted') and a.get('media_file_basename') in filenames
    ]


def format_submission_time(iso: str) -> str:
    from datetime import datetime, timezone
    try:
        dt = datetime.fromisoformat(iso.replace('Z', '+00:00'))
        return dt.astimezone(timezone.utc).strftime('%d %b %Y, %H:%M')
    except Exception:
        return iso
